sunbeam-python/sunbeam/core/watcher.py

```diff
@@ -0,0 +1,213 @@
+        goal=WORKLOAD_BALANCING_GOAL_NAME,
+        strategy=WORKLOAD_BALANCING_STRATEGY_NAME,
+    )
+    return template
+
+
+def get_enable_maintenance_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(ENABLE_MAINTENANCE_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_host_maintenance_audit_template(client=client)
+    return template
+
+
+def get_workload_balancing_audit_template(
+    client: watcher_client.Client,
+) -> watcher.AuditTemplate:
+    try:
+        template = client.audit_template.get(WORKLOAD_BALANCING_AUDIT_TEMPLATE_NAME)
+    except NotFound:
+        template = _create_workload_balancing_audit_template(client=client)
+    return template
+
+
+@tenacity.retry(
+    reraise=True,
+    stop=tenacity.stop_after_delay(WAIT_TIMEOUT),
+    wait=tenacity.wait_fixed(WAIT_SLEEP_INTERVAL),
+)
+def _wait_resource_in_target_state(
+    client: watcher_client.Client,
+    resource_name: str,
+    resource_uuid: str,
+    states: list[str] = ["SUCCEEDED", "FAILED"],
+) -> watcher.Audit:
+    src = getattr(client, resource_name).get(resource_uuid)
+    if src.state not in states:
+        raise SunbeamException(f"{resource_name} {resource_uuid} not in target state")
+    return src
+
+
+def create_audit(
+    client: watcher_client.Client,
+    template: watcher.AuditTemplate,
+    audit_type: str = "ONESHOT",
+    parameters: dict[str, Any] = {},
+) -> watcher.Audit:
+    audit = client.audit.create(
+        audit_template_uuid=template.uuid,
+        audit_type=audit_type,
+        parameters=parameters,
+    )
+    audit_details = _wait_resource_in_target_state(
+        client=client,
+        resource_name="audit",
+        resource_uuid=audit.uuid,
+    )
+
+    if audit_details.state == "SUCCEEDED":
+        LOG.debug(f"Create Watcher audit {audit.uuid} successfully")
+    else:
+        LOG.debug(f"Create Watcher audit {audit.uuid} failed")
+        raise SunbeamException(
+            f"Create watcher audit failed, template: {template.name}"
+        )
+
+    _check_audit_plans_recommended(client=client, audit=audit)
+    return audit
+
+
+def _check_audit_plans_recommended(client: watcher_client.Client, audit: watcher.Audit):
+    action_plans = client.action_plan.list(audit=audit.uuid)
+    # Verify all the action_plan's state is RECOMMENDED.
+    # In case there is not action been generated, the action plan state
+    # will be SUCCEEDED at the beginning.
+    if not all(plan.state in ["RECOMMENDED", "SUCCEEDED"] for plan in action_plans):
+        raise SunbeamException(
+            f"Not all action plan for audit({audit.uuid}) is RECOMMENDED"
+        )
+
+
+def get_actions(
+    client: watcher_client.Client, audit: watcher.Audit
+) -> list[watcher.Action]:
+    """Get list of actions by audit."""
+    return client.action.list(audit=audit.uuid, detail=True)
```

Comment:
Add limit=0 to get all the action plans [1]

[1] https://opendev.org/openstack/python-watcherclient/src/commit/edbc62c502e0448555548a718ba3fc7af530984b/watcherclient/v1/action.py#L42